Fix ChainStorageError after a reorg with new block

hansieodendaal opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Fix ChainStorageError after a reorg with new block

- Fixed a ChainStorageError that occurs due to invalid transactions remaining in the mempool when the 1st block is requested for mining via GRPC directly after a chain reorg.
- Added a failing cucumber test for the above condition (fixed now): `Mempool clearing out invalid transactions after a reorg`.
- Minor cucumber improvements.
- ~~Fixed flaky test `tx_broadcast_protocol_submit_success`~~

## Motivation and Context
See above

## How Has This Been Tested?
Cucumber test `Mempool clearing out invalid transactions after a reorg`

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [X] I'm merging against the `tari-script` branch.
* [X] I have squashed my commits into a single commit.

philipr-za commented on base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs:
```diff
@@ -176,36 +176,71 @@ impl UnconfirmedPool {
     }
 
+    /// Remove all current mempool transactions from the UnconfirmedPoolStorage, returning that which have been removed
+    pub fn clear_out_all_mempool_transactions(&mut self) -> Vec<Arc<Transaction>> {
```
I think the more canonically Rust term for this process would be `drain`

philipr-za commented on base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs:
```diff
@@ -176,36 +176,71 @@ impl UnconfirmedPool {
 
+    /// Remove all current mempool transactions from the UnconfirmedPoolStorage, returning that which have been removed
+    pub fn clear_out_all_mempool_transactions(&mut self) -> Vec<Arc<Transaction>> {
+        let mut mempool_txs: Vec<Arc<Transaction>> = Vec::new();
```
Performance nit: when you know how long the vector is going to be you should ideally use `Vec::with_capacity(N)`

hansieodendaal replied:
Code changed as suggested below, not applicable anymore.

philipr-za commented on base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs:
```diff
@@ -176,36 +176,71 @@ impl UnconfirmedPool {
+                tx_key.get_signature().to_hex()
+            );
+            mempool_txs.push(ptx.transaction.clone());
+            self.txs_by_priority.remove(&ptx.priority);
```
I might be misunderstanding something about these collections but it feels unnecessary to do this one entry at a time. You could just drain the hashmap values into the return vector and clear these two collections. Why do this one at a time?

hansieodendaal replied:
Tried to do it as a batch, but had some runtime issues. Will look at it again.

stringhandler replied:
Agree with Phil here. This is a really complicated way of clearing out two variables. 

```
let mempool_txs = self.txs_by_signature.drain(..).map(....).collect();
self.txs_by_priority = Vec::new();
return mempool_txs;
```

hansieodendaal replied:
Done

philipr-za commented on base_layer/wallet/tests/transaction_service/transaction_protocols.rs:
```diff
@@ -255,7 +255,7 @@ async fn tx_broadcast_protocol_submit_success() {
     let join_handle = task::spawn(protocol.execute());
 
-    // Fails because there is no transaqction in the database to be broadcast
+    // Fails because there is no transaction in the database to be broadcast
```
I believe that SW's PR here https://github.com/tari-project/tari/pull/2899 also addresses this issue

hansieodendaal replied:
Noted.

hansieodendaal replied:
Fix removed

Commit:
Fix ChainStorageError after a reorg with new block

- Fixed a ChainStorageError that occurs due to invalid transactions remaining in
the mempool when the 1st block is requested for mining via GRPC directly after
a chain reorg.
- Added a failing cucumber test for the above condition.
- Minor cucumber improvements.
- Updated the `tari-crypto` version number dependency.

stringhandler commented on base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs:
```diff
@@ -176,36 +176,71 @@ impl UnconfirmedPool {
+
     /// Remove all published transactions from the UnconfirmedPool and discard all double spend transactions.
-    /// Returns a list of all transactions that were removed the unconfirmed pool as a result of appearing in the block.
+    /// Returns a list of all transactions that were removed from the unconfirmed pool as a result of appearing in the
```
nit: This comment is out of date. This method does not return anything

stringhandler reviewed (approved):
Some changes needed to be followed up in further PRs, but this can go in as is for now

hansieodendaal commented:
> Some changes needed to be followed up in further PRs, but this can go in as is for now

All changes effected.

hansieodendaal commented:
See #2969 where the failing `Scenario: Pruned mode reorg past horizon` test has been made predictable.
